ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/LfValueTranslation.scala

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import anorm.NamedParameter
+import com.daml.caching
+import com.daml.ledger.api.v1.value.{Record => ApiRecord, Value => ApiValue}
+import com.daml.ledger.EventId
+import com.daml.ledger.api.v1.event.{CreatedEvent, ExercisedEvent}
+import com.daml.metrics.Metrics
+import com.daml.platform.participant.util.LfEngineToApi
+import com.daml.platform.store.dao.events.{Value => LfValue}
+import com.daml.platform.store.serialization.ValueSerializer
+
+final class LfValueTranslation(val cache: LfValueTranslation.Cache) {
+
+  private def cantSerialize(attribute: String, forContract: ContractId): String =
+    s"Cannot serialize $attribute for ${forContract.coid}"
+
+  private def serializeCreateArgOrThrow(contractId: ContractId, arg: LfValue): Array[Byte] =
+    ValueSerializer.serializeValue(
+      value = arg,
+      errorContext = cantSerialize(attribute = "create argument", forContract = contractId),
+    )
+
+  private def serializeCreateArgOrThrow(c: Create): Array[Byte] =
+    serializeCreateArgOrThrow(c.coid, c.coinst.arg)
+
+  private def serializeNullableKeyOrThrow(c: Create): Option[Array[Byte]] =
+    c.key.map(
+      k =>
+        ValueSerializer.serializeValue(
+          value = k.key,
+          errorContext = cantSerialize(attribute = "key", forContract = c.coid),
+      )
+    )
+
+  private def serializeExerciseArgOrThrow(e: Exercise): Array[Byte] =
+    ValueSerializer.serializeValue(
+      value = e.chosenValue,
+      errorContext = cantSerialize(attribute = "exercise argument", forContract = e.targetCoid),
+    )
+
+  private def serializeNullableExerciseResultOrThrow(e: Exercise): Option[Array[Byte]] =
+    e.exerciseResult.map(
+      exerciseResult =>
+        ValueSerializer.serializeValue(
+          value = exerciseResult,
+          errorContext = cantSerialize(attribute = "exercise result", forContract = e.targetCoid),
+      )
+    )
+
+  // Doesn't go through caching, for now caching is limited to events
+  def serialize(contractId: ContractId, createArgument: LfValue): Vector[NamedParameter] =
```

Comment:
We can avoid wrapping in Vector in this case.

Reply:
Address https://github.com/digital-asset/daml/pull/6052/commits/d3c67a7f69d52e9397de6b9a03adf4e945f85bb6